### ISSUE TITLE
PE Version?

### ISSUE THREAD

**Issue by kentikzlo** [ProtocolSupport/ProtocolSupport#304]
When you update pe protocol support?

**Comment by Shevchik** [comment]
Maybe next year.

**Comment by Creeperface01** [comment]
It would be awesome!

**Comment by MrPowerGamerBR** [comment]
Now we are on "next year" 😋 

(ProtocolSupportPE was miles better than any PE server software)

**Comment by Shevchik** [comment]
Yes, but currently the main goal is to port ProtocolSupport to GlowStone, and then i can start thinking about implementing PE.

**Comment by Creeperface01** [comment]
Ok :)

**Comment by DaMatrix** [comment]
@Shevchik If there's demand for this i can work on it too. I'm sick of feature requests on my server for things that aren't in pocketmine :/

**Comment by MrPowerGamerBR** [comment]
I also can help by... I don't know, saying "whoa you are awesome keep up the good work!"

(Okay, I can do more useful things than that)

**Comment by Shevchik** [comment]
Okay first we need someone familiar with PE itself, so we know how things work.
Then we need someone with IDA knowledge because i don't have enough time to both decompile mcpe and code.
And we need some testers too.

**Comment by MrPowerGamerBR** [comment]
@Shevchik yeah, getting a MCPE protocol documentation is difficult because there isn't a good documentation like Minecraft PC. However I doubt you will need to decompile MCPE to know the codes.

MiNET has a good auto generated protocol if you need https://github.com/NiclasOlofsson/MiNET/blob/master/src/MiNET/MiNET/Net/MCPE%20Protocol%20Documentation.md

For protocol you can also see how PocketMine-MP (or Nukkit) implements packets as a "documentation"

Before you get your hair pulled out because you can't figure out why the fuck the client doesn't login (I had this issue...), here is the packet order for the login process:

Player sends a LoginPacket (0x01)
Server sends a PlayStatusPacket (0x02) with status 0
Server sends a StartGamePacket (0x0c)
Server sends a PlayStatusPacket (0x02) with status 3

This a very very very basic handshake, but at least allows you to get the client to login (as in: the player is connected to a world and not stuck on the "Locating server..."/"Building terrain..." screen)

If you need help with the protocol, you can ask @NiclasOlofsson, he is the creator of MiNET and he is probably the person with most knowledge about packets, heck, probably more than Mojang itself!

Also, if "testers" is "someone who can't help with the project at all except for adding some packet translators and etc etc etc but can test with Minecraft Windows 10 Edition + Minecraft Pocket Edition", then count on me!

**Comment by Shevchik** [comment]
The problem is that most fields are still undocumented, or even worse they are actually wrong and it doesn't crash only by pure luck. Or worse due to mcpe inventory being retarded you never know how actually things work, see inventory open which actually pulls inventory info from real time of packet info, or chunk packet which actually expects all tiles nbt to be sent. Which means i still have to decompile mcpe.

**Comment by kentikzlo** [comment]
@Shevchik I can test new protocolsupport pe, i have a pe project with big online and I ready to support you with money, if it wil be successful 👍

**Comment by NiclasOlofsson** [comment]
@Shevchik There are very few fields that remain undocumented, and even fewer are wrong.

And what lacks in documentation is well made up by assistance by the dev community that deals with the protocols. It's not a big community (maybe 5-10 ppl), but a lot of collaboration going on. Some are good at IDA, some have clients, some just eat bytes for breakfast. But they all help.

But if you believe you need to disassemble MCPE to get the NBT out, yeah .. then maybe you shouldn't waste your time on doing PE.

The only real difficult thing about PE is dealing with encryption. Not because it is difficult to understand, but because it is difficult to get the proper tool-set to deal with it.

**Comment by DaMatrix** [comment]
Is encryption really needed? To my knowledge it's only needed for players who are logged in with xbox live, and even xbox players can connect to servers without encryption enabled, a decent login plugin would suffice.

**Comment by MrPowerGamerBR** [comment]
@DaMatrix when I was creating my MCPE server from scratch the client (nor my server) never sent any encryption request/response, I'm not using Xbox Live login so I don't know if it is the same thing if you are using Xbox Live.

But if you aren't using Xbox Live (or doesn't care if the client is using Xbox Live login) you can just not send the "ServerToClientHandshake" and never enable encryption, I think the client doesn't care about that.

For a minimal handshake only three packets are used: LoginPacket, PlayStatusPacket and StartGamePacket.

**Comment by NiclasOlofsson** [comment]
@DaMatrix No, encryption is not mandatory. And you can use XBL without encryption.

**Comment by MrPowerGamerBR** [comment]
@DaMatrix just for science I enabled Xbox Live on MCPE just to see what was going to happen when connecting to my "made from scratch" server, and yep, the client doesn't care at all about encryption.

So don't send "ServerToClientHandshake" after the LoginPacket and then you don't need to care about encryption at all.

**Comment by Shevchik** [comment]
Well, maybe something changed, i haven't looked at mcpe for a while. Is mcpe inventory still mostly clientside (so you have to guess what actually moved by searching items), or mojang finally fixed this?

**Comment by Creeperface01** [comment]
Yes its fixed.

**Comment by MrPowerGamerBR** [comment]
@Shevchik they probably fixed that when they released Minecraft Windows 10
Edition. (Which has the Java PC GUI for inventory)

And moving items in a MCPE server while using MCW10 client works like what
you would expect in PC Minecraft.


Em 22 de jan de 2017 12:44, "Shevchik" <notifications@github.com> escreveu:

Well, maybe something changed, i haven't looked at mcpe for a while. Is
mcpe inventory still mostly clientside (so you have to guess what actually
moved by searching items), or mojang finally fixed this?

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/ProtocolSupport/ProtocolSupport/issues/304#issuecomment-274334937>,
or mute the thread

**Comment by Shevchik** [comment]
Well, this makes things easier. I'm hoping that we finish glowstone support next week, and then i can start working on PE.

**Comment by Shevchik** [comment]
Also we need someone with deep knowledge of RakNet, because there are some issues in it which i don't know how to solve. For example:
1. How to disconnect client with custom message when error happened during the initial connection request (Encapsulated layer should not be available yet).
2. That happened long time ago, so maybe it is fixed, but there was a problems with ordering index which sometimes skipped ids. (So i had to order packets by seq index)
3. What to do when indexes overflow?

**Comment by DaMatrix** [comment]
1. https://github.com/Nukkit/Nukkit/blob/master/src/main/java/cn/nukkit/Player.java#L1907

**Comment by Shevchik** [comment]
That's an encapsulated layer packet which is not avilable at RakNet handshake.

**Comment by MrPowerGamerBR** [comment]
@DaMatrix I think Nukkit never kicks a player before processing the player's LoginPacket.

Well, @MarfGamer has his own RakNet version for Java (that he created to use in his own MCPE server software) and he still maintains it, so maybe he knows those questions.

**Comment by MrPowerGamerBR** [comment]
@Shevchik this should be useful for you ;) https://github.com/sel-project/sel-utils ~ https://github.com/sel-project/sel-utils/blob/master/doc/pocket/100.md

(Or maybe not... anyway, maybe it will be useful for someone else)

**Comment by DaMatrix** [comment]
Aside from the lack of comments, that format reminds me of wiki.vg's MCPC protocol documentation. looks very useful.

**Comment by MrPowerGamerBR** [comment]
@DaMatrix for me the useful part is the "constants", a generated protocol already exists (MiNET auto generated documentation), but it doesn't includes the valid constants for that packet.

Example: The play status packet is ID 2 and has a unsigned int as a value, but what are the valid values for that unsigned int? MiNET's documentation only has a few comments about constants for some packets, but that documentation has for every packet (okay, I didn't see every packet, but it seems it has for all)

**Comment by MrPowerGamerBR** [comment]
The repository also has auto generated Minecraft Pocket Edition packets in Java (and other languages): https://github.com/sel-project/sel-utils/tree/master/src/java/sul/protocol/pocket100/play

**Comment by MrPowerGamerBR** [comment]
@DaMatrix another very useful thing in this documentation is the Metadata entity flags and MCPE attributes, I never seen something like that before in any other MCPE protocol documentation.

However I also found out that there are some packets missing from that documentation. (As far as I know: The send map item packet is missing (the one used to send the map item data information))

**Comment by MrPowerGamerBR** [comment]
Just a small correction about something I said before:

To successfully connect a MCPE client to a server, you need to send the following packets:
Player sends a LoginPacket (0x01)
Server sends a PlayStatusPacket (0x02) with status 0
Server sends a ResourcePacksInfoPacket (0x07)
Server sends a StartGamePacket (0x0c)
Server sends a PlayStatusPacket (0x02) with status 3

Yes, the "ResourcePacksInfoPacket" is required, if not sent the client will never join the world.

**Comment by MarfGamer** [comment]
@MrPowerGamerBR I know this thread is an old topic and that this probably has nothing to do with it, but could you DM me on Twitter? I followed you. I would love to have you on the BeetRoot-MP team as you seem to have important info about the protocol that I have been trying to get a hold of forever now and you seem that you would be a valuable addition to the team!

P.S. The message about the login sequence, thank you so much for posting that. I have spent weeks trying to get the client to join the world, without that I couldn't have get clients spawning (even if its just a void, haha)!

**Comment by MrPowerGamerBR** [comment]
@MarfGamer but... didn't you already add me to your organization before? You added me when I asked when you would start working with your server software in the JRakNet repo.

(And my protocol knowledge isn't that good, I only read https://github.com/sel-project/sel-utils and how Nukkit does things)

**Comment by MarfGamer** [comment]
Oh ok, I didn't know I had added you. Whoops

**Comment by Shevchik** [comment]
Okay, just need to fix book issues and then we can start.

**Comment by Shevchik** [comment]
I guess i fixed almost all issues in ps 1.11, so i'm planning to release it tomorrow and start working on MCPE. I will use a JRakLib for a while, and will switch to my own raknet over netty udp implementation later.

**Comment by MrPowerGamerBR** [comment]
@Shevchik if you need help, I'm here. (I'm not that experienced with packets and things, but I learned a lot about the Pocket protocol in this month and how to implement them, I even added an experimental Pocket protocol to Glowstone! https://github.com/PocketDreams/SequinLand-Glowstone (which... isn't finished but you can login, walk around, view other players, view other entities, etc))

And JRakNet works fine, MarfGamer fixed a lot of issues that I reported and now it is possible to use it for a Minecraft Pocket Edition server.

**Comment by Shevchik** [comment]
The problem is that half of the ps internal things requires netty as network engine.

**Comment by DaMatrix** [comment]
Just so everyone watching this thread knows, MCPE is now being developed my @Shevchik [here](https://github.com/ProtocolSupport/ProtocolSupport/tree/mcpenew).

**Comment by yangm97** [comment]
Is anyone hosting compiled builds of the experimental branch or should we build ourselves?

**Comment by 7kasper** [comment]
It isn't really stable yet, but there are some cool features already :-)

Because of that I assume that no builds are released, but building is quite easy:
1. Clone in git or download the zip file from the mcpenew ([mcpenew](https://github.com/ProtocolSupport/ProtocolSupport/tree/mcpenew) branch.
2. When downloaded, unzip.
3. Open the directory with all the files in cmd or bash. (On windows you can do file -> open commandprompt)
4. type: gradlew jar
5. After gradlew is finished building your compiled ProtocoSupport.ljar is located in the target folder.

**Comment by yangm97** [comment]
I can see a lot of stuff has been added already. Awesome!

I also would like to know if it will be possible to use it alongside bungeecord, authme and authmebridge. I know I could put dragon proxy in front of a bungeecord network and be done with it, but since this project is implemented as a plugin on the server side, how could one implement such setup? Would that require native PE support from bungeecord? A ProtocolSupport plugin for BungeeCord? An authmebridge plugin for a bungeecord PE alternative? Something else?

Thank you all.

**Comment by gdude2002** [comment]
The short answer is no - Bungee does not support MCPE.

That said, PE doesn't really have any kind of authentication, so you would be able to push PC clients through Bungee and allow the PE clients to connect directly (on another port).

PE also has a special packet that is used to signal clients to connect to another server, so you could in theory achieve the same effect, but yeah, there's definitely some work to go before this kind of thing becomes possible.

**Comment by yangm97** [comment]
> The short answer is no - Bungee does not support MCPE.

I know that.

>  PE also has a special packet that is used to signal clients to connect to another server…

It only works if both servers are “public”. A whole bungeecord network can be private, with the only exception being the bungeecord proxy itself.

Looks like there are two main problems: authentication and teleportation. I think a ProtocolSupport plugin for bungeecord could nail both, by reusing the existing bungeecord an authme infrastructure.

Also, is this issue the best place to talk about this?

**Comment by Shevchik** [comment]
We have discord channel where we usually talk about such things.

**Comment by MrPowerGamerBR** [comment]
@yangm97 

> I think a ProtocolSupport plugin for bungeecord could nail both,

ProtocolSupportPE needs to have access to the internal entity storage to get the absolute position of entities (due to how MCPE protocol works) so ProtocolSupportPE for BungeeCord is impossible. (okay, not really impossible, but still, it is too hard)

If you get a bit creative you can create a server network with ProtocolSupportPE (I have some ideas for that) but I doubt you can just drag and drop ProtocolSupportPE to your PC server and it would magically work.

**Comment by yangm97** [comment]
> We have discord channel where we usually talk about such things.

Oops, missed the discord badge.

> If you get a bit creative you can create a server network with ProtocolSupportPE

I started thinking about that too. A pocketmine or whatever server running an auth plugin compatible with authme database would handle authentication and then we could re-create the bungeecord network by using some bungee alternative for PE. 

But I assume that would not work with current multi-server portal implementations that rely on bungeecord API.

**Comment by DaMatrix** [comment]
There already is an alternative to Bungee for PE, although I'm not sure if it's still mantained. I believe it was under https://github.com/iTXTech/Nemisys . I'm not sure how well it works though, as i haven't used it in a LOOOOOOOOONG time.

**Comment by CubeFromSpace** [comment]
@damatrix that's their old version:
https://github.com/iTXTech/Genisys

**Comment by caspervanneck** [comment]
@CubeFromSpace  that is the software not the transfer software

**Comment by DaMatrix** [comment]
I was talking about Nemisis, Genisys is something completely different.

**Comment by CubeFromSpace** [comment]
Oh, I was confusing with something different

**Comment by jfmherokiller** [comment]
thank you for working on this but i must add that on mac you will have issues compiling it because of missing certificates so it will fail to get the required prerequisites.

**Comment by gdude2002** [comment]
@jfmherokiller The latter (which is my server) is using LetsEncrypt. LetsEncrypt was added to the default Java keystore in one of the Java 8 releases - that's what you need.

**Comment by CubeFromSpace** [comment]
I think after @DaMatrix and (crap I forgot the head dev's name) declared DragonProxy and TheDragonetTeam dead, they showed me this repo and said something about ProtocolSupporrPE

**Comment by TheDiamondYT1** [comment]
@Shiny-andKittens That head dev would be @DefinitelyEvil. I worked on, it but kinda fucked it up 😂

**Comment by DaMatrix** [comment]
PSPE is on hold for a while, as we wait for PC 1.12 and PE to add server-sided inventories.

**Comment by yangm97** [comment]
Sorry to be that guy but… what’s the status of this?

**Comment by gdude2002** [comment]
MCPE beta 1.2 is out, so work can start again.

**Comment by Asuna-Yuuki** [comment]
Yay

**Comment by gdude2002** [comment]
Please use the reactions system instead of one-word comments.

**Comment by BrettplayMC** [comment]
What server versions will be supported on this branch?

**Comment by gdude2002** [comment]
Whatever the latest versions are at the time of release.

**Comment by MrPowerGamerBR** [comment]
Only the latest PE version + every version ProtocolSupport already supports

Em 21 de set de 2017 10:30, "BrettplayMC" <notifications@github.com>
escreveu:

**Comment by Miegl** [comment]
Just compiled the mcpenew branch, tried to use it with 1.12.2 spigot server, but for some reason the server didn't show up in MCPE (tried 1.1.4 and 1.2.0) and manually joining the server didn't work too. What am I missing here?

**Comment by MrPowerGamerBR** [comment]
@Miegl it isn't ready for production, the server port isn't the default port (19132) used in MCPE, you can figure what port ProtocolSupportPE uses by poking around in the source code.

(this was a deliberate choice to avoid users from using ProtocolSupportPE before it was finished)

**Comment by Miegl** [comment]
@MrPowerGamerBR oh thanks

**Comment by Miegl** [comment]
@MrPowerGamerBR well actually, it works very well! Thanks for your work, amazing stuff!

**Comment by Miegl** [comment]
The only major bug that I encountered is weird camera position flipping left and right when trying to walk towards stairs or slabs

**Comment by MrPowerGamerBR** [comment]
@Miegl known bug

**Comment by BrettplayMC** [comment]
Oh so only the latest Spigot? Rip.

**Comment by Shevchik** [comment]
latest version of platform we support (spigot and glowstone currently)

**Comment by FurmigaHumana** [comment]
the new "better together update" allows all versions to play together (except java version obviously), will this update make anything easier for a future protocolsuport connecting all versions?

**Comment by 7kasper** [comment]
@FurmigaHumana Yes..
Because of the better together update Pocket Edition, Windows 10 Edition, Gear VR Edition, Apple TV Edition, Fire TV Edition, Xbox One Edition and Nintendo Switch Edition are now all the _same_ game. That means if this stuff hits the fan, all those versions will be (technically) able to play on a java server. This still excludes some versions (like the xbox 360, new nintendo 3ds, etc) which we will most likely _never_  support and Console versions will require a proxy to join a server, as only local and official Mojang partner servers can be joined.

We were also waiting for this update because serverside inventories our now finally a thing, which was a requirement to basically play minecraft. The problem currently however is that most systems in the Bedrock engine (eg. Minecraft Pocket Edition, but on all platforms) are sooo different from the PC version, that they are hard to port.

**Comment by MrPowerGamerBR** [comment]
tl;dr: We are going to support only Minecraft Pocket Edition and Minecraft Windows 10 Edition until Microsoft stops being a dumbass and starts supporting external servers in the other versions.

Also it will take a long time until you will be able to use it on your Minecraft server.

**Comment by Wqrld** [comment]
bump they support it now right?

**Comment by MrPowerGamerBR** [comment]
No they don't.

**Comment by Wqrld** [comment]
but how do you than join mineplex and lifeboat etc, and i remember there already being a proxy that lets pe people join mc servers but it was stopped

**Comment by MrPowerGamerBR** [comment]
You can join external servers in MCW10E/MCPE, my response above was about Minecraft Xbox Edition and Minecraft Switch Edition, both doesn't support external servers aside from the four partner servers.

**Comment by Wqrld** [comment]
since most people are still pe, this would not really be a problem right?

**Comment by Donovan22** [comment]
Is there currently a fix for the random freezes? I tried all of the branches. Am I missing something?

**Comment by EOT3000** [comment]
when will you add bungeecord support?

**Comment by gdude2002** [comment]
You mean [PSBungee](https://github.com/ProtocolSupport/ProtocolSupportBungee)?

**Comment by WesleyVanNeck** [comment]
@EOT3000  psbungee got pe support but you have to eneble it and you need it find on your own by sniffing in the code

**Comment by WesleyVanNeck** [comment]
But am off topic now so I will stop now

**Comment by EOT3000** [comment]
oh sorry

**Comment by alexchandel** [comment]
Is there a jenkins build for the latest mcpenew?  I'd like to test it.

**Comment by person4268** [comment]
As far as I know, no, but I have a build if you want it. 


[ProtocolSupport.zip](https://github.com/ProtocolSupport/ProtocolSupport/files/2991872/ProtocolSupport.zip)

> 
> 
> Is there a jenkins build for the latest mcpenew? I'd like to test it.

**Comment by Heath123** [comment]
If you want an alternative try https://github.com/Geyser/GeyserMC
It's incomplete but it's mostly playable, and it has standalone and plugin versions

**Comment by OrangeKale64** [comment]
well no one ever uses old versions of pe though

**Comment by LegacyGamerHD** [comment]
@OrangeKale64 if you have patient they are working on it at least the geyser team

**Comment by OrangeKale64** [comment]
@LegacyGamerHD ???

**Comment by LegacyGamerHD** [comment]
> @LegacyGamerHD ???

It's already out geyser 1.16 so don't bother me

**Comment by OrangeKale64** [comment]
@LegacyGamerHD 🤦🏻

**Comment by Heath123** [comment]
> @LegacyGamerHD 🤦🏻

Why are you sending a facepalm emoji?

**Comment by OrangeKale64** [comment]
Because he dosn't get that I mean I don't know a single person who uses an old version of mcpe

**Comment by LegacyGamerHD** [comment]
> Because he doesn't get that I mean "I don't know A SINGLE person who uses an old version of bedrock edition."

I dont think you get me geyser is updated to 1.16.x and i get you but you dont get me...

**Comment by FurmigaHumana** [comment]
@Shevchik @7kasper should this thread be closed? It has clearly gone off-topic and I dont see any progress made on PE versions (at least there is nothing public).. with Geyser around i dont see a reason why waste resources on this..

**Comment by LegacyGamerHD** [comment]
> @Shevchik @7kasper should this thread be closed? It has clearly gone off-topic and I dont see any progress made on PE versions (at least there is nothing public).. with Geyser around i dont see a reason why waste resources on this..

Do you feel offended  by it so no pspe is dead as af

**Comment by FurmigaHumana** [comment]
> > @Shevchik @7kasper should this thread be closed? It has clearly gone off-topic and I dont see any progress made on PE versions (at least there is nothing public).. with Geyser around i dont see a reason why waste resources on this..
> 
> Do you feel offended by it so no pspe is dead as af

just wanting to keep it professional...

**Comment by LegacyGamerHD** [comment]
> > > @Shevchik @7kasper should this thread be closed? It has clearly gone off-topic and I dont see any progress made on PE versions (at least there is nothing public).. with Geyser around i dont see a reason why waste resources on this..
> > 
> > 
> > Do you feel offended by it so no pspe is dead as af
> 
> just wanting to keep it professional...

Why the hell would we need this professional 1 its about minecraft 2 its community  project so I dont see why you wand it professional

**Comment by FurmigaHumana** [comment]
> > > > @Shevchik @7kasper should this thread be closed? It has clearly gone off-topic and I dont see any progress made on PE versions (at least there is nothing public).. with Geyser around i dont see a reason why waste resources on this..
> > > 
> > > 
> > > Do you feel offended by it so no pspe is dead as af
> > 
> > 
> > just wanting to keep it professional...
> 
> Why the hell would we need this professional 1 its about minecraft 2 its community project so I dont see why you wand it professional

This project is something very few people could do, even more for free, so yes it need to be respected because people are giving their own personal time for it.. if devs are cool with this so alright... but for me as a full time developer who also happens to give time for the open comunity i would never accept such behavior for something i owe nothing to

**Comment by LegacyGamerHD** [comment]
> > > > > @Shevchik @7kasper should this thread be closed? It has clearly gone off-topic and I dont see any progress made on PE versions (at least there is nothing public).. with Geyser around i dont see a reason why waste resources on this..
> > > > 
> > > > 
> > > > Do you feel offended by it so no pspe is dead as af
> > > 
> > > 
> > > just wanting to keep it professional...
> > 
> > 
> > Why the hell would we need this professional 1 its about minecraft 2 its community project so I dont see why you wand it professional
> 
> This project is something very few people could do, even more for free, so yes it need to be respected because people are giving their own personal time for it.. if devs are cool with this so alright... but for me as a full time developer who also happens to give time for the open comunity i would never accept such behavior for something i owe nothing to

you dont even remotly make sense i repsect shevchick also its only chevchick working on it 7kasper didnt contribute for a very long time also you aint a dev here also i did notting wrong so point out @FurmigaHumana  also why do mix with it to make more problems ? and you aint a dev there is notting on your github profile even if you work on a difrent platform why do even mix with that

**Comment by Heath123** [comment]
Calm down, someone just asked if it should be closed or not

**Comment by LegacyGamerHD** [comment]
> Calm down, someone just asked if it should be closed or not

ok

**Comment by 7kasper** [comment]
Any discussion about this topic should be done in our discord. I see no reason to keep this thread unlocked in this way. 
If we drop our plans or PSPE is ready and gets implemented into the main PS line 🐌 , this issue / feature request will be closed.
Yes I have not contributed in a while. No I also don't know the exact plans. But I think there is some hope left and there are reasons why we have not closed this thread yet. Take care.

**Comment by Shevchik** [comment]
We don't have plans on dropping the plan to implement pe support. It might take a while to get some alpha version released, but you can use alternatives (such as Geyser) for now.
